### PR TITLE
[models]: update master PT add checks for each reco model and xorresponding test

### PR DESCRIPTION
**Pull request by felixdittrich92**
This PR:

- apply suggestions for MASTER PT @frgfm 
- add check each recognition model if train mode that targets != None and corresponding test

Any feedback is welcome :hugs: 

Inference again evaluated pipe and single word crop:

```
Line(
            (words): [
              Word(value='des', confidence=0.86),
              Word(value='Hausanschus', confidence=0.47),
              Word(value='und', confidence=0.83),
              Word(value='Keleraumes', confidence=0.39),
              Word(value='erhalten', confidence=0.75),
            ]
          ),
```
results are really impressive :+1:

### CODE REVIEW

**Comment by felixdittrich92** [comment]
Note: failing tests (link to external link currently not available) has nothing to do with this PR :)

**Review by charlesmindee — APPROVED** [review]
Thanks, LGTM

File: doctr/models/recognition/crnn/pytorch.py [hunk 1]
```diff
@@ -194,6 +194,9 @@ def forward(
         return_preds: bool = False,
     ) -> Dict[str, Any]:
 
+        if self.training and target is None:
+            raise ValueError('Need to provide labels during training for teacher forcing')
```
Comment (frgfm) [inline doctr/models/recognition/crnn/pytorch.py:198]: I think the message should be different here. Teacher forcing is specific to the SAR :sweat_smile:

File: doctr/models/recognition/crnn/tensorflow.py [hunk 2]
```diff
@@ -195,6 +195,9 @@ def call(
         **kwargs: Any,
     ) -> Dict[str, Any]:
 
+        if kwargs.get('training', False) and target is None:
+            raise ValueError('Need to provide labels during training for teacher forcing')
```
Comment (frgfm) [inline doctr/models/recognition/crnn/tensorflow.py:199]: same here

File: doctr/models/recognition/master/pytorch.py [hunk 3]
```diff
@@ -176,6 +176,9 @@ def forward(
 
         out: Dict[str, Any] = {}
 
+        if self.training and target is None:
+            raise ValueError('Need to provide labels during training for teacher forcing')
```
Comment (frgfm) [inline doctr/models/recognition/master/pytorch.py:180]: same here

File: doctr/models/recognition/master/pytorch.py [hunk 4]
```diff
@@ -213,7 +216,7 @@ def decode(self, encoded: torch.Tensor) -> torch.Tensor:
         b = encoded.size(0)
 
         # Padding symbol + SOS at the beginning
-        ys = torch.zeros((b, self.max_length), dtype=torch.long, device=encoded.device) * (self.vocab_size + 2)  # pad
+        ys = torch.ones((b, self.max_length), dtype=torch.long, device=encoded.device) * (self.vocab_size + 2)  # pad
```
Comment (frgfm) [inline doctr/models/recognition/master/pytorch.py:219]: Let's avoid the multiplication:
```python
ys = torch.full((b, self.max_length), self.vocab_size + 2, dtype=torch.long, device=encoded.device) # pad
```

File: doctr/models/recognition/transformer/pytorch.py [hunk 5]
```diff
@@ -59,25 +59,25 @@ def scaled_dot_product_attention(
...
+            nn.Dropout(p=dropout),
+            nn.Linear(ffd, d_model),
+        )
 
     def forward(self, x: torch.Tensor) -> torch.Tensor:
-        x = torch.relu(self.first_linear(x))
-        x = self.dropout(x)
-        x = self.sec_linear(x)
-        return x
+        return self.feed_forward(x)
```
Comment (frgfm) [inline doctr/models/recognition/transformer/pytorch.py:72]: This is wrapping a Sequential, we can remove that overhead:
```python
class PositionwiseFeedForward(nn.Sequential):
    """ Position-wise Feed-Forward Network """
    def __init__(self, d_model: int, ffd: int, dropout: float = 0.1) -> None:
        super().__init__(
            nn.Linear(d_model, ffd),
            nn.ReLU(),
            nn.Dropout(p=dropout),
            nn.Linear(ffd, d_model),
        )
```

and you won't even need to override the forward method

**Review by frgfm — COMMENTED** [review]
Sorry, late review but I think there were some possible improvements :)

File: doctr/models/recognition/crnn/pytorch.py (hunk 1 above)
Comment (felixdittrich92) [reply doctr/models/recognition/crnn/pytorch.py:198]: Oh yes i agree without a similar (bigger) model or RNN/LSTM this makes no sense ..

File: doctr/models/recognition/master/pytorch.py (hunk 4 above)
Comment (felixdittrich92) [reply doctr/models/recognition/master/pytorch.py:219]: Let me check this again, if i remember there was a problem with full but i will check it again :)
